proposals/gc/Overview.md

```diff
@@ -1 +1,696 @@
+
+Packed fields require special load/store instructions:
+```
+(load_field_packed_s $s $a (...))
+(load_field_packed_u $s $a (...))
+(store_field_packed $s $a (...) (...))
+(load_elem_packed_s $s $a (...))
+(load_elem_packed_u $s $a (...))
+(store_elem_packed $s $a (...) (...))
+```
+
+
+### Mutability
+
+Fields and elements can either be immutable or *mutable*:
+```
+(type $s (struct (field $a (mut i32)) (field $b i32)))
+(type $a (array (mut i32)))
+```
+Store operators are only valid when targeting a mutable field or element.
+
+Immutability is needed to enable the safe and efficient [subtyping](#subtyping), especially as needed for the [objects](#objects-and-mehtod-tables) use case.
+
+
+### Nullability
+
+By default references cannot be null,
+avoiding any runtime overhead for null checks when using them.
+
+Nullable references are available as separate types called `optref`.
+
+TODO: Design a casting operator that avoids the need for control-flow sensitive typing.
+
+
+### Defaultability
+
+Most value types, including all numeric types and nullable references are *defaultable*, which means that they have 0/null as a default value.
+Other reference types are not defaultable.
+
+Certain restrictions apply to non-defaultable types:
+
+* Local declarations of non-defaultable type must have an initializer.
+* Allocations of aggregates with non-defaultable fields or elements must have initializers.
+
+Objects whose members all have _mutable_ and _defaultable_ type may be allocated without initializers:
+```
+(type $s (struct (field $a (mut i32)) (field (mut (ref $s)))))
+(type $a (array (mut f32)))
+
+(new_default_struct $s)
+(new_default_array $a (i32.const 100))
+```
+
+
+### Sharing
+
+TODO: Distinguish types safe to share between threads in the type system.
+
+
+## Other Reference Types
+
+### Universal Type
+
+The type `anyref` can hold references of any reference type.
+It can be formed via [up casts](#casting),
+and the original type can be recovered via [down casts](#casting).
+
+
+### Foreign References
+
+A new built-in value type called `foreignref` represents opaque pointers to objects on the _embedder_'s heap.
+
+There are no operations to manipulate foreign references, but by passing them as parameters or results of exorted Wasm functions, embedder references (such as DOM objects) can safely be stored in or round-trip through Wasm code.
+```
+(type $s (struct (field $a i32) (field $x foreignref))
+
+(func (export "f") (param $x foreignref)
+  ...
+)
+```
+
+
+### Function References
+
+References can also be formed to function types, thereby introducing the notion of _typed function pointer_.
+
+Function references can be called through `call_ref` instruction:
+```
+(type $t (func (param i32))
+
+(func $f (param $x (ref $t))
+  (call_ref (i32.const 5) (get_local $x))
+)
+```
+Unlike `call_indirect`, this instruction is statically typed and does not involve any runtime check.
```

Comment:
Wouldn't homogeneous tables simply have a plain function type as element type?

I can see that with function references we might want to add instructions to access tables directly, but that in turn seems independent of homogeneous tables.

Trying to unravel the dependencies. It seems that homogeneous function tables are unrelated to GC types. The only feature through which tables (both homegeneous and heterogeneous) might interact with this proposal would be the instructions you mention.